JS: Refactor the XSS / Client-side-url queries

erik-krogh opened the pull request:
This PR replaces https://github.com/github/codeql/pull/6632 

See [this comment by Asger](https://github.com/github/codeql/pull/6632#discussion_r705269103) for what this PR tries to fix.  

This is a deep fix where flow-labels are used to hopefully correctly model how url sinks can cause XSS.  

There are now 3 flow labels in the XSS query:   
- TaintedUrlSuffix: a URL where the attacker only controls a suffix.
- Taint: a tainted value where the attacker controls part of the value.
- PrefixLabel: a tainted value where the attacker controls the prefix 

--- 

All tests pass on every commit (the expected files are updated on each commit).  
This should help in doing a commit-by-commit review. 

--- 

[An evaluation looks good](https://github.com/github/codeql-dca-main/tree/data/PR-8304-0-javascript/reports) (I think).  
I like the new alerts, and I don't miss the alerts that got removed.    
(I didn't triage the `xss-through-dom` results). 

(Sidenote: If you think a user-controlled `src=".."` attribute on an `<img/>` tag is safe, then checkout SVG parsing in browsers).  

There is a small performance penalty, but I think we can live with that.

asgerf reviewed (commented):
Great! I'm very happy to see progress on this issue.

[This result](https://github.com/ampproject/amphtml/blob/5f5a1ad80281ffdd0022b84469552d1ac702fc40/examples/analytics-iframe-transport-remote-frame.html#L39) seems to be a FP due to not recognizing the call `url.startsWith('https://3p.ampproject.net/')` as a sanitizer guard.

The URL sinks could be further divided into cases where
- tainted scheme leads to XSS (`<a href=...>`). Use the `Prefix` label.
- tainted hostname leads to XSS  (`<script src=...>`). Currently no flow label for this.

We could add another flow label for the hostname URL sinks, and use sanitizers from `UrlConcatenation` for that label. Would be interesting to see how much it affects performance.

esbena commented on javascript/ql/lib/semmle/javascript/security/dataflow/ClientSideUrlRedirectCustomizations.qll:
```diff
@@ -19,7 +19,12 @@ module ClientSideUrlRedirect {
-  abstract class Sink extends DataFlow::Node { }
+  abstract class Sink extends DataFlow::Node {
+    /** Holds if the sink can execute JavaScript code in the current context. */
+    predicate isXSSSink() {
```
(ql-for-ql: four upper-case characters in a row are suspicious...)

erik-krogh replied:
> (ql-for-ql: four upper-case characters in a row are suspicious...)

[That rabbit hole was deep.](https://github.com/github/codeql/pull/8323)

erik-krogh commented:
> [This result](https://github.com/ampproject/amphtml/blob/5f5a1ad80281ffdd0022b84469552d1ac702fc40/examples/analytics-iframe-transport-remote-frame.html#L39) seems to be a FP due to not recognizing the call `url.startsWith('https://3p.ampproject.net/')` as a sanitizer guard. 

It's actually from `url` being a global variable, and that causes us to not recognize the use within the `if` as sanitized.   
The sanitizer works if you place the entire thing inside a function.

erik-krogh commented on javascript/ql/lib/semmle/javascript/security/dataflow/DomBasedXssQuery.qll:
```diff
@@ -78,6 +74,32 @@ class JQueryHtmlOrSelectorInjectionConfiguration extends TaintTracking::Configur
+    )
+    or
+    // we assume that `.join()` calls have a prefix, and thus block the prefix label.
+    node = any(DataFlow::MethodCallNode call | call.getMethodName() = "join") and
```
> join calls are also modelled as string concatenations. Is the above case not enough?

Yes join calls are modeled as string concats, but only when `join` is called with the empty string.   
So e.g. `[taint, "constant"].join("/")` is not modeled as a string concatenation.   

So no, the above case is not enough.

erik-krogh commented on javascript/ql/lib/semmle/javascript/security/TaintedUrlSuffix.qll:
```diff
@@ -26,6 +26,17 @@ module TaintedUrlSuffix {
+  ClientSideRemoteFlowSource source() {
+    result.getKind().isFragment()
+    or
+    result.getKind().isQuery()
```
I still think we need to include all sources of kind `url`.  
In the QLDoc for `ClientSideRemoteFlowKind::isUrl` is states: `the untrusted part of the URL is prefixed by trusted data`.

erik-krogh commented:
> The URL sinks could be further divided into cases where

If possible I would prefer waiting with that. 
Your suggested change requires changing `HostnameSanitizerGuard` so that it's somehow configurable which labels should be sanitized (possibly make the class abstract and instantiate the class in all the queries that use it).

asgerf commented:
> It's actually from url being a global variable, and that causes us to not recognize the use within the if as sanitized.
The sanitizer works if you place the entire thing inside a function.

Ah, in that case I'm OK with going ahead with this.

Some tests are currently red, but otherwise LGTM.

erik-krogh commented:
The latest CI failure was due to a semantic merge conflict with https://github.com/github/codeql/pull/8323  

I rebased on `main` and updated updated a name-reference to a non-deprecated name.
